riscv_isac/coverage.py

```diff
@@ -1065,6 +1146,16 @@ def write_fn_csr_comb_covpt(csr_reg):
                                         simd_val_unpack(value['val_comb'], op_width, "rs2", rs2_val, lcls)
                                     instr_vars.update(lcls)
                                     for coverpoints in value['val_comb']:
+                                        pattern_imm = r'\bimm_val\b'
+                                        pattern_rs1 = r'\brs1_val\b'
+                                        pattern_rs2 = r'\brs2_val\b'
+                                        match_imm = bool(re.search(pattern_imm, coverpoints))
+                                        match_rs1 = bool(re.search(pattern_rs1, coverpoints))
+                                        match_rs2 = bool(re.search(pattern_rs2, coverpoints))
+                                        if (instr_vars['rs2_val'] == None) and match_rs2:
```

Comment:
I don't think the coverpoints for instructions without `rs2_val` and `imm_val` should be skipped. This should not be causing problems unless the variable is being indexed somewhere. Can you provide an example which causes this error to arise?

Reply:
Please consider the example of following coverpoint, the instruction of interest will be picked by the **riscv-isac** using the following `mnemonics`.
```objective-c
  mnemonics:
    sw: 0
    csrrw: 0
    csrrs: 0
    csrrc: 0
    lw: 0
```
Also consider the following coverpoint written under the val_comb label. So, for the instructions like lw and sw which will obviously of interest because of the mnemonics will properly work for this coverpoint.
```objective-c
  val_comb:
    get_pte_prop("${U_bit}",get_addr("begin_signature"), rs1_val + imm_val, get_addr("rvtest_slvl1_pg_tbl")) == 1: 0

```
But, for the instructions which will have mnemonics like `csrrw`,`csrrs`,`csrrc` in them, will actually have no `imm_val` in the **instr_vars** dictionary and also the rs2_val will be of None Type. So, we can't perform any operations on them.

-------------------------------------------------------------------
One solution that I have tried to implement is to simply skip the instructions without imm_val and rs2_val.

The other solution can be to write the coverpoints in such a manner that they skip the instructions which do not contain the imm_val and rs2_val. But, for that I will have to change the following initialization of `imm_val` in the **Instruction_Object.py**.

https://github.com/riscv-software-src/riscv-isac/blob/db7504037ee611aa5b7e0b40e5207ae4871ef574/riscv_isac/InstructionObject.py#L159-L160

This will be changed such that the value of imm_val from the **c_sail.py** will always be added in the **instr_vars** even if its None so we can use the coverpoint properly with imm_val and rs2_val being of **None Type**

-----------------------------------------------------------------
So, for the example of coverpoint, we may write something like:
```objective-c
  val_comb:
    (get_pte_prop("${U_bit}",get_addr("begin_signature"), rs1_val + imm_val, get_addr("rvtest_slvl1_pg_tbl")) == 1) if imm_val is not None else False: 0

```

Reply:
My understanding of this issue is that if they are not skipped, and are used as part of a computation (e.g. being indexed is one example), then they will fail. This can be fixed in one of two ways: explicitly skipped in the code that is defined for the instruction format, or modifying the .cgf files to effectively skip them. There are tradeoffs both ways, but the preference is to minimize the work, especially where there are future modifications to coverage (but not the formats, so explicitly skipping in the code was viewed as needing less work in the future.

Reply:
The grouping under `mnemonics` were only intended towards instructions which share coverpoints and hence the underlying architectural variables. If the underlying variables are not the same, they should not be sharing the same coverpoints and should be placed in different covergroups.

Reply:
Yes, you are right, this was actually a short term solution, so I have removed this feature, so, in the coverpoints where we are targeting some specific instruction, we can do that by using the mnemonic flag(https://github.com/riscv-software-src/riscv-isac/pull/80#issue-1992987280) instead of changing the coverage.py for every instruction that does not have rs2_val and imm_val. So, the coverpoint look something like:
```
val_comb:
    mnemonic == ('sw' or 'lw') and (get_pte_prop("r",get_addr("begin_signature"), rs1_val + imm_val, get_addr("rvtest_slvl1_pg_tbl")) == 1): 0
```